crates/oxc_query/src/edges.rs

```diff
@@ -0,0 +1,2169 @@
+        resolve_neighbors_with(contexts, |v| {
+            let attr = v.as_jsx_attribute().unwrap_or_else(|| {
+                panic!("expected to have a jsxattribute vertex, instead have: {v:#?}")
+            });
+            let Some(attr_value) = &attr.value else { return Box::new(std::iter::empty()) };
+            Box::new(
+                std::iter::once(match attr_value {
+                    JSXAttributeValue::ExpressionContainer(expr) => match &expr.expression {
+                        oxc_ast::ast::JSXExpression::Expression(expr) => Some(Vertex::from(expr)),
+                        oxc_ast::ast::JSXExpression::EmptyExpression(_) => None,
+                    },
+                    JSXAttributeValue::Fragment(_)
+                    | JSXAttributeValue::StringLiteral(_)
+                    | JSXAttributeValue::Element(_) => None,
+                })
+                .flatten(),
+            )
+        })
+    }
+
+    pub(super) fn value_as_url<'a, 'b: 'a>(
+        contexts: ContextIterator<'a, Vertex<'b>>,
+        _resolve_info: &ResolveEdgeInfo,
+    ) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+        resolve_neighbors_with(contexts, |v| {
+            #[allow(clippy::option_if_let_else)]
+            if let Some(url) = Vertex::make_url(v.as_jsx_attribute().unwrap_or_else(|| {
+                panic!("expected to have a jsxattribute vertex, instead have: {v:#?}")
+            })) {
+                Box::new(std::iter::once(url))
+            } else {
+                Box::new(std::iter::empty())
+            }
+        })
+    }
+}
+
+pub(super) fn resolve_jsxelement_edge<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    edge_name: &str,
+    _parameters: &EdgeParameters,
+    resolve_info: &ResolveEdgeInfo,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+    match edge_name {
+        "child_element" => jsxelement::child_element(contexts, resolve_info),
+        "child_expression_container" => {
+            jsxelement::child_expression_container(contexts, resolve_info)
+        }
+        "child_fragment" => jsxelement::child_fragment(contexts, resolve_info),
+        "child_spread" => jsxelement::child_spread(contexts, resolve_info),
+        "child_text" => jsxelement::child_text(contexts, resolve_info),
+        "opening_element" => jsxelement::opening_element(contexts, resolve_info),
+        "span" => jsxelement::span(contexts, resolve_info),
+        _ => {
+            unreachable!("attempted to resolve unexpected edge '{edge_name}' on type 'JSXElement'")
+        }
+    }
+}
+
+mod jsxelement {
+    use oxc_ast::ast::JSXChild;
+    use trustfall::provider::{
+        resolve_neighbors_with, ContextIterator, ContextOutcomeIterator, ResolveEdgeInfo,
+        VertexIterator,
+    };
+
+    use super::{super::vertex::Vertex, get_span};
+    use crate::vertex::JSXElementVertex;
+
+    macro_rules! child_edge_iter {
```

Comment:
another macro to simplify to unify the implementation of 4 functions, is this worth it or too confusing?

Reply:
not my favorite but not the worst either